src/model/openthaigpt_pretraining_model/data_wrapper.py

```diff
@@ -1,37 +1,176 @@
 import torch
 from torch.utils.data import IterableDataset
+from datasets import Dataset, load_dataset, load_from_disk
+import os
+from re import findall
+from tqdm import tqdm
+from transformers import (
+    AutoTokenizer,
+    LlamaTokenizer,
+)
+from ..datasets.constants import SPLIT_TRAIN, SPLIT_VAL
 
-HF_TOKENIZER_INPUT_IDS_NAME = "input_ids"
 
+class TokenizedDataset:
+    def __init__(
+        self,
+        mode: str,
+        tokenizer: str,
+        max_tokens: int = 2048,
+        save_path: str = "./",
+        chunk_size: int = 1024 * 1024,
+        batch_size: int = 10000,
+        num_proc: int = 16,
+        dataset_name: str = "oscar",
+        dataset_dir: str = "unshuffled_deduplicated_th",
+    ):
+        if len(findall("llama", tokenizer)):
+            self.tokenizer = LlamaTokenizer.from_pretrained(tokenizer)
+            self.tokenizer.pad_token = "<pad>"
+            self.tokenizer.add_special_tokens({"pad_token": "<pad>"})
+        else:
+            self.tokenizer = AutoTokenizer.from_pretrained(tokenizer)
+
+        self.mode = mode
+        self.max_tokens = max_tokens
+        self.save_path = save_path
+        self.chunk_size = chunk_size
+        self.batch_size = batch_size
+        self.num_proc = num_proc
+        if mode == "val":
+            self.data_set = load_dataset(
+                dataset_name,
+                dataset_dir,
+                split=SPLIT_VAL,
+            )
+        elif mode == "train":
+            self.data_set = load_dataset(
+                dataset_name,
+                dataset_dir,
+                split=SPLIT_TRAIN,
+            )
+        else:
+            raise NotImplementedError("only support Train,Val")
+        self.num_shards = (len(self.data_set) + chunk_size - 1) // chunk_size
 
-class DatasetWrapper(IterableDataset):
```

Comment:
Why you remove this? We are using this class [here](https://github.com/OpenThaiGPT/openthaigpt-pretraining/blob/1e1c481f78be18552bb154fdf66d78e761560817/src/model/openthaigpt_pretraining_model/optimizers/lion/utils.py#L25).

Reply:
sorry about that

Reply:
I restored it now